### DISCUSSION TITLE
Best Practice for Using GetX Dependency Injection Without GetMaterialApp

### DISCUSSION THREAD

**Discussion by LeonManolo** [jonataslaw/getx#3070] [Q&A]
Hello,

I am planning to conduct a performance comparison among various state management systems within Flutter, aiming to maintain a consistent app structure throughout the testing process. One of my goals is to explore the most efficient way to utilize GetX's dependency injection features without relying on `GetMaterialApp`. 

In the context of using Bloc, I'm familiar with wrapping the `MaterialApp` widget with `BlocProvider` to ensure global access to the Bloc across the application. This approach allows for a neat integration with the Flutter framework while maintaining the ability to inject dependencies as needed.

However, when it comes to GetX, I understand that `GetMaterialApp` is typically recommended for leveraging the full suite of GetX's capabilities, including routing, state management, and dependency injection. My question is: What would be the best practice for implementing GetX dependency injection in a Flutter app without using `GetMaterialApp`? I'm looking for a method that would allow me to maintain a similar level of global accessibility and ease of integration as the `BlocProvider` does for Bloc, but within the constraints of using GetX for state management and dependency injection.

My primary intention is to ensure that the app structure remains consistent while testing different state management systems, to achieve a fair comparison in terms of performance and usability. Any guidance or recommendations on how to best approach this with GetX would be greatly appreciated.

Thank you in advance for your assistance.

**Comment by jonataslaw** [comment]
Initially, GetMaterialApp should only be used for routes, snackbars, and modals. If you only use dependency injection and state management, it is unnecessary.

Second, if you don't use routes, auto-dispose won't happen, unless you use one of two alternatives:

1. Use a Widget that injects the dependency (GetBuilder, Bind, etc.).
2. Use a StatefulWidget to inject and remove the dependency
3. You can even create a specific widget that injects the dependency in onInit and removes it in dispose.

GetX is flexible (sometimes too much), and you can do it however you see fit. All the 3 ways above will work perfectly, without any problems.

**Reply by FullstackWEB-developer** [reply]
As a developer who loves Getx, I am currently upgrading my project to BLOC due to pressure from some who argue that [large-scale projects should use BLOC](https://youtu.be/zlIgy4es5Ts).
The main issue here was the removal of `GetMaterialApp` raised by @LeonManolo .
But now there seems to be no obstacle.
@jonataslaw  So does this mean that using dependency injection using `GetBuilder` and `Bind` in a widget will achieve the same results as using `get_it`?

Again, I really like the clean and responsive pattern of Getx, so I don't understand the need for BLOC yet.
